packages/ERTP/src/mathHelpers/setMathHelpers.js

```diff
@@ -98,6 +165,44 @@ const setMathHelpers = harden({
     const leftElemNotInRight = leftElem => !hasElement(rightBuckets, leftElem);
     return harden(left.filter(leftElemNotInRight));
   },
+  doFind: (left, searchParameters) => {
+    let matchNotFound = false;
+    let arrayOfArrays;
+    try {
+      arrayOfArrays = searchParameters.map(searchRecord => {
+        const getBucketKey = makeGetBucketKeyBasedOnSearchRecord(searchRecord);
+        const leftBuckets = makeBuckets(left, getBucketKey);
+        const bucketKey = getBucketKey(searchRecord);
+        const maybeMatches = leftBuckets.get(bucketKey);
+        if (maybeMatches === undefined) {
+          matchNotFound = true;
+          throw Error('match was not found');
+        }
+        const matches = maybeMatches.filter(maybeMatch =>
+          // for every key that exists in searchRecord, the value for
+          // that key in maybeMatch must be sameStructure
+          Object.entries(searchRecord).every(([key, value]) =>
+            sameStructure(maybeMatch[key], value),
+          ),
+        );
+        if (matches.length <= 0) {
+          matchNotFound = true;
+          throw Error('match was not found');
+        }
+        return matches;
+      });
+    } catch (err) {
+      if (matchNotFound) {
+        // At least one searchRecord had no match
+        return identity;
```

Comment:
Unfortunately, for JavaScript this is probably the best way to do this. In tc39 I had suggested generalizing `break <label>;` so that you could break to a label in a lexically enclosing function, giving us a non-erroneous non-local exit. It would make JS more expressive and arguably simpler at the same time, since it "simply" removes a restriction that would no longer need to be explained. Ah well.